File: Ni/Ni/Services/PostService.cs
using Microsoft.Azure.Storage;
using Microsoft.Azure.Storage.Blob;
using Ni.Core.Entities;
using Ni.Core.Requests;
using Ni.Core.Responses;
using Ni.Core.Services;
using Ni.Repositories;
using System;
using System.Collections.Generic;
using System.Linq;

namespace Ni.Services
{
    public class PostService : IPostService
    {
        private IAuthKeyRepository _authKeyRepository;
        private IPostRepository _postRepository;
        private ITagRepository _tagRepository;
        private IUserRepository _userRepository;
        private ICategoryRepository _categoryRepository;

        public PostService(IPostRepository postRepository, IAuthKeyRepository authKeyRepository,
            ITagRepository tagRepository, IUserRepository userRepository, ICategoryRepository categoryRepository)
        {
            _postRepository = postRepository;
            _authKeyRepository = authKeyRepository;
            _tagRepository = tagRepository;
            _userRepository = userRepository;
            _categoryRepository = categoryRepository;
        }

        public GenericResponse AddPost(AddPostRequest request)
        {
            GenericResponse response = new GenericResponse();
            response.Errors = new List<string>();
            bool validation = _authKeyRepository.Validate(request.RequesterId, request.AuthKey);
            if (!validation)
            {
                response.StatusCode = 400;
                response.Errors.Add("You do not have access");
            }
            else
            {
                int postId = _postRepository.AddPost(request.RequesterId, request.Title, request.Content);
                foreach (var tag in request.Tags)
                {
                    _tagRepository.AddToPost(postId, tag);
                }
                response.StatusCode = 200;
                string connString = "DefaultEndpointsProtocol=https;AccountName=timefunctionnip8ca1;AccountKey=yexSIebMMLZTojXbWS6QSvSOEoxXIqQW3l9oNjfpjdwVRrQKYgMQj1cgrrU7aHnVygp+TQPRsGoeq1UkWaxkUw==;EndpointSuffix=core.windows.net";
                CloudStorageAccount.TryParse(connString, out var storageAccount);
                var cloudBlobClient = storageAccount.CreateCloudBlobClient();
                var container = cloudBlobClient.GetContainerReference("files");
                var cloudBlockBlob = container.GetBlockBlobReference(postId + ".png");
                byte[] imageBytes = Convert.FromBase64String(request.Image);
                cloudBlockBlob.UploadFromByteArray(imageBytes, 0, imageBytes.Length);
            }
            return response;
        }

        public GetPostsResponse GetAll(GetAllPostsRequest request)
        {
            GetPostsResponse response = new GetPostsResponse();
            response.StatusCode = 200;
            response.Errors = new List<string>();
            response.Posts = new List<PostDTO>();
            var posts = _postRepository.GetAll();
            foreach (var post in posts)
            {
                var tags = new List<string>();
                var tagEntities = _tagRepository.GetByPostId(post.Id);
                foreach (var tag in tagEntities)
                {
                    tags.Add(tag.Content);
                }
                PostDTO postWithTags = new PostDTO()
                {
                    Post = post,
                    Tags = tags,
                    AuthorUsername = _userRepository.GetUserById(post.UserId).Username,
                };
                response.Posts.Add(postWithTags);
            }
            return response;
        }

        public GetPostsResponse GetAllByCategory(GetAllPostsByCategoryRequest request)
        {
            GetPostsResponse response = new GetPostsResponse();
            response.StatusCode = 200;
            response.Errors = new List<string>();
            response.Posts = new List<PostDTO>();
            var cat = _categoryRepository.GetCategoryByUrl(request.CategoryURL);
            if(cat == null)
            {
                response.StatusCode = 404;
                response.Errors.Add("Category not found");
                return response;
            }
            var posts = _postRepository.GetByCategory(cat.Id);
            foreach (var post in posts)
            {
                var tags = new List<string>();
                var tagEntities = _tagRepository.GetByPostId(post.Id);
                foreach (var tag in tagEntities)
                {
                    tags.Add(tag.Content);
                }
                PostDTO postWithTags = new PostDTO()
                {
                    Post = post,
                    Tags = tags,
                    AuthorUsername = _userRepository.GetUserById(post.UserId).Username,
                };
                response.Posts.Add(postWithTags);
            }
            return response;
        }

        public GetPostsResponse GetLatest(GetLatestPostsRequest request)
        {
            GetPostsResponse response = new GetPostsResponse();
            response.StatusCode = 200;
            response.Errors = new List<string>();
            response.Posts = new List<PostDTO>();
            var posts = _postRepository.GetLatest(request.Start,request.Count);
            foreach (var post in posts)
            {
                var tags = new List<string>();
                var tagEntities = _tagRepository.GetByPostId(post.Id);
                foreach (var tag in tagEntities)
                {
                    tags.Add(tag.Content);
                }
                PostDTO postWithTags = new PostDTO()
                {
                    Post = post,
                    Tags = tags,
                    AuthorUsername = _userRepository.GetUserById(post.UserId).Username,
                };
                response.Posts.Add(postWithTags);
            }
            return response;
        }

        public GetPostResponse GetPostById(GetPostByIdRequest request)
        {
            GetPostResponse response = new GetPostResponse();
            response.StatusCode = 200;
            response.Errors = new List<string>();
            var tags = new List<string>();
            var tagEntities = _tagRepository.GetByPostId(request.PostId);
            foreach (var tag in tagEntities)
            {
                tags.Add(tag.Content);
            }
            var post = _postRepository.GetPostById(request.PostId);
            response.Post = new PostDTO()
            {
                Post = post,
                Tags = tags,
                AuthorUsername = _userRepository.GetUserById(post.UserId).Username,
            };
            return response;
        }
    }
}

File: Ni/Web/ClientApp/src/app/services/post/post.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {ServerService} from '../server/server.service';
import {GetPostsResponse} from '../../models/posts/getPostsResponse';
import {GetLatestPostsRequest} from '../../models/posts/getLatestPostsRequest';
import {GetPostByIdRequest} from '../../models/posts/getPostByIdRequest';
import {GenericResponse} from '../../models/genericResponse';
import {AddPostRequest} from '../../models/posts/addPostRequest';
import {GetPostResponse} from '../../models/posts/getPostResponse';
import {GetAllPostsByCategoryRequest} from '../../models/posts/getAllPostsByCategoryRequest';

@Injectable({
    providedIn: 'root'
})
export class PostService {

    private postsUrl = 'Post';
    private postIdUrl = 'post/Id/';
    private postLatestUrl = 'post/Latest/';
    private postsAllUrl = 'post/All/';
    private postsAllCategoryUrl = 'post/Category/';

    constructor(public serverService: ServerService,
                public http: HttpClient) {
        this.serverService = serverService;
        this.http = http;
    }

    async addPost(requesterId: number, authKey: string, title: string, image: string, content: string, tags: string[]) {
        const request = new AddPostRequest();
        request.requesterId = requesterId;
        request.authKey = authKey;
        request.title = title;
        request.image = image;
        request.content = content;
        request.tags = tags;
        return this.http.post<GenericResponse>(this.serverService.mainServerUrl + this.postsUrl, request,
            {headers: this.serverService.requestHeaders}).toPromise();
    }

    async getAllPosts() {
        return this.http.get<GetPostsResponse>(this.serverService.mainServerUrl + this.postsAllUrl, {
            headers: this.serverService.requestHeaders
        }).toPromise();
    }
    async getAllPostsByCategory(category: string) {
        const request = new GetAllPostsByCategoryRequest();
        request.categoryURL = category;
        return this.http.get<GetPostsResponse>(this.serverService.mainServerUrl + this.postsAllCategoryUrl, {
            params: request as any,
            headers: this.serverService.requestHeaders
        }).toPromise();
    }

    async getlatestPosts(start: number, count: number) {
        const request = new GetLatestPostsRequest();
        request.start = start;
        request.count = count;
        return this.http.get<GetPostsResponse>(this.serverService.mainServerUrl + this.postLatestUrl, {
            params: request as any,
            headers: this.serverService.requestHeaders
        }).toPromise();
    }

    async getPostById(id: number) {
        const request = new GetPostByIdRequest();
        request.postId = id;
        return this.http.get<GetPostResponse>(this.serverService.mainServerUrl + this.postIdUrl, {
            params: request as any,
            headers: this.serverService.requestHeaders
        }).toPromise();
    }
}
